go/vt/vttablet/tabletmanager/rpc_external_reparent.go

```diff
@@ -193,51 +200,117 @@ func (agent *ActionAgent) finalizeTabletExternallyReparented(ctx context.Context
+		errs.RecordError(err)
+	}
+
+	if errs.HasErrors() {
+		return errs.Error()
+	}
+
 	if !topoproto.TabletAliasIsZero(oldMasterAlias) {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
+		// Tell the old master to re-read its tablet record and change its state.
+		// We don't need to put error into errs if this fails, but we need to wait
+		// for it to make sure that old master tablet is not stuck in the MASTER
+		// state.
+		tmc := tmclient.NewTabletManagerClient()
+		if err := tmc.RefreshState(ctx, oldMasterTablet); err != nil {
+			log.Warningf("Error calling RefreshState on old master %v: %v", topoproto.TabletAliasString(oldMasterTablet.Alias), err)
+		}
+	}
 
-			// Tell the old master to re-read its tablet record and change its state.
-			// We don't need to put error into errs if this fails, but we need to wait
-			// for it to make sure that old master tablet is not stuck in the MASTER
-			// state.
-			tmc := tmclient.NewTabletManagerClient()
-			if err := tmc.RefreshState(ctx, oldMasterTablet); err != nil {
-				log.Warningf("Error calling RefreshState on old master %v: %v", topoproto.TabletAliasString(oldMasterTablet.Alias), err)
-			}
-		}()
+	tabletMap, err := agent.TopoServer.GetTabletMapForShard(ctx, masterTablet.Keyspace, masterTablet.Shard)
+	if err != nil {
+		log.Errorf("ignoring error %v from GetTabletMapForShard so that we can process any partial results", err)
 	}
 
-	wg.Wait()
+	if len(tabletMap) > 0 {
+		// make the channel buffer big enough that it doesn't block senders
+		tabletsToRefresh := make(chan topodatapb.Tablet, len(tabletMap))
+		// update any other tablets claiming to be MASTER also to REPLICA
+		for _, tabletInfo := range tabletMap {
+			alias := tabletInfo.Tablet.Alias
+			if !topoproto.TabletAliasEqual(alias, agent.TabletAlias) && !topoproto.TabletAliasEqual(alias, oldMasterAlias) && tabletInfo.Tablet.Type == topodatapb.TabletType_MASTER {
+				log.Infof("finalizeTabletExternallyReparented: updating tablet record for another old master: %v", alias)
+				wg.Add(1)
+				go func(alias *topodatapb.TabletAlias) {
+					defer wg.Done()
+					var err error
+					tab, err := agent.TopoServer.UpdateTabletFields(ctx, alias,
+						func(tablet *topodatapb.Tablet) error {
+							if tablet.Type == topodatapb.TabletType_MASTER {
+								tablet.Type = topodatapb.TabletType_REPLICA
+							}
+							return nil
+						})
+					if err != nil {
+						errs.RecordError(err)
+						return
+					}
+					// tab will be nil if no update was needed
+					if tab != nil {
+						tabletsToRefresh <- *tab
+					}
+				}(alias)
+			}
+		}
+		// Wait for the tablet records to be updated. At that point, any rebuild will
+		// see the new master, so we're ready to mark the reparent as done in the
+		// global shard record.
+		wg.Wait()
+		// we waited for all goroutines to complete, so now close the channel
+		close(tabletsToRefresh)
+		if errs.HasErrors() {
+			return errs.Error()
+		}
+
+		for tab := range tabletsToRefresh {
```

Comment:
We can combine these last two steps to simplify the code now that we're not doing the shard update in the middle. Just call RefreshState after UpdateTabletFields in each per-tablet goroutine launched above.